### PR TITLE
Add validation to whitelist form

### LINKED ISSUE #1684: Add validation to site whitelisting form on the options page
The form should accept domain-like inputs only. URLs should be automatically corrected to domains.

Following up on https://github.com/EFForg/privacybadger/issues/1607#issuecomment-331289343.

### PR DESCRIPTION
**Pull request by n-sayenko**
Only accept URL and domain-like user inputs. 
If adding URL, automatically convert to domain. 
If not a valid URL, display alert message.
fixes #1684

### CODE REVIEW

**Comment by ghostwords** [comment]
Hello and welcome! I'll have a better look tomorrow, just a couple of things for now.

I edited your description to properly auto-close the parent issue (the word "fixes" has to be immediately next to the issue number).

Also, there is no need to add the new translation messages to the non-default (`en_US`) locale since this is something we can take care of automatically after we sign off on everything else the PR. We should probably document this fact somewhere.

**Comment by ghostwords** [comment]
Hi, sorry, it's taking me longer to get back to this. I wonder if we should reuse `extractHostFromURL` and `URI` from [basedomain.js](https://github.com/EFForg/privacybadger/blob/080b1b5cbe2c9b056e835f666ffbb47bcd1142c6/src/lib/basedomain.js) to validate input and convert URIs to domains, to avoid introducing regexes/duplicating functionality.

**Comment by n-sayenko** [comment]
That's okay! I agree, it's probably better to use the functions from `basedomain.js` . I didn't look through the files carefully enough. I can change it.

**Comment by n-sayenko** [comment]
Hey, I'm realizing that `URI` and `extractHostFromURL` only work with URLs that start with https:// and http://, so the user wouldn't be able to input `domain.com` for example.  Is it okay to rework those functions a little bit?

**Comment by ghostwords** [comment]
Maybe we could do something like the following?
```
if userInput doesn't start with http
  prefix userInput with http://

if userInput doesn't end with "/"
  postfix userInput with "/"

try {
  url = new URI(userInput)
} catch (e) {
  // I think we have to use try/catch to handle URI() errors ...
  userInput is invalid
}

if userInput is valid
  domain = url.host
```

**Comment by ghostwords** [comment]
Before calling the `URI` constructor, you could try prefixing user input with "http://" if user input doesn't already start with "http://" or "https://". And, to avoid failing due to trivial formatting differences, also postfixing user input with "/", if user input doesn't already end with "/".

Anything that is or could be a valid http(s) URL will pass, anything else will fail. If we have a valid URL, we get its domain for saving.

**Comment by n-sayenko** [comment]
Just realized with my last commit that now inputting gibberish will work, because i'm adding "http://"  and "/" to it, so `URL`  doesn't seem to ever return an error.  But urls will still output domains. I'll rework this some time tomorrow, sorry for the delay.

**Comment by ghostwords** [comment]
That might be OK ... The main issue we want to avoid is people entering URLs when we require domains. I don't think we need to care much beyond that.

File: src/js/options.js [hunk 1]
```diff
@@ -285,17 +285,40 @@ function getOriginsArray(filterText) {
   return Object.keys(originCache).filter(containsFilterText);
 }
 
+// if url doesn't have http in front, add 
+function makeValidURL(url) {
+  var match = url.match(/^(?:https?:)/);
```
Comment (ghostwords) [inline src/js/options.js]: Can we simplify with `url.startsWith`?

File: src/js/options.js [hunk 2]
```diff
@@ -285,17 +285,40 @@ function getOriginsArray(filterText) {
...
+// if url doesn't have http in front, add 
+function makeValidURL(url) {
+  var match = url.match(/^(?:https?:)/); 
+  var endURL = url;
+
+  if(!match) {
+    endURL =  "http://" + url;
+  }
+
+  if(endURL.slice(-1) !== "/") {
```
Comment (ghostwords) [inline src/js/options.js]: Could do `url.endsWith` for readability.

**Comment by n-sayenko** [comment]
In that case, I can't find any case where `URI` returns an error, should I delete the `invalid_domain` message?

**Comment by ghostwords** [comment]
Not everything validates ... for example, `http:example`. We should also check that `validURL.host` is truthy, since we could have a "valid" URL with an empty host, apparently (for example, `@`).

Could you fix the lint warnings in https://travis-ci.org/EFForg/privacybadger/jobs/291822206#L610-L621?

**Comment by ghostwords** [comment]
@n-sayenko Would you like me to finish this pull request? No rush, just checking in. Let me know if you'd like to make the fixes and look this over again yourself.

It would also be nice to take the validation logic and wrap it into a function so that we could add a few unit tests to verify what should pass, what should fail, and how certain input (URLs) should get transformed (domains).

**Comment by n-sayenko** [comment]
@ghostwords  Hi, really sorry for the delay! I don't think I'm going to have time for this until this weekend. If you're okay with waiting until then I can make the fixes.

If not, feel free to take over and finish it up.

**Comment by ghostwords** [comment]
Yep, that sounds good. Happy to wait and review when you are ready.

**Comment by n-sayenko** [comment]
Hey did you mean put the try/catch into a separate function? or did you want `makeValidUrl` and the try/catch in the same separate function?

**Comment by ghostwords** [comment]
I think what you got it is good, although checking `validURL.host` should happen inside the validation function where we should also display our error message if we don't have a host.

**Comment by ghostwords** [comment]
Thank you! Merged.

I made a few followup tweaks over several commits you can review [here](https://github.com/EFForg/privacybadger/compare/80be109063be9043b34dd471b77c2881dd8aaf56%5E...55424abd55dd0826554a70308cca3468d2f6543c). Let me know if you have any questions.

**Pull request merged**